docs/20240515-zedtokens.md

```diff
@@ -0,0 +1,46 @@
+# Consistency with ZedTokens
+
+## Overview
+
+In SpiceDB, resources are arranged in a graph where different services may be contributing to that graph. For example, permissions-api itself manages roles and role bindings, while tenant-api manages tenants. Resources are updated over NATS request/reply.
+
+SpiceDB breaks updates down into quantized revisions of configurable duration, meaning ACL updates within a given revision window are not visible by default until the next revision begins. Using ZedTokens, however, clients can request data at least as fresh as a given exact point in time. This allows for clients to control the consistency they need for a given SpiceDB request on a per-request basis.
+
+In general, services expect that immediately after updating relationships in permissions-api, they can make authorization decisions based on data at least as fresh as when they made those updates.
+
+## Goals
+
+This proposal is meant to achieve the following goals:
+
+* Permissions checks made on a resource after it is updated must use data at least as fresh as the last update to that resource
+
+## Non-goals
+
+This proposal is not meant to achieve the following goals:
+
+* Updates to the graph are globally and immediately propagated
+
+## Proposed solution
+
+To mitigate this issue, permissions-api can be updated to use a table in CRDB to populate ZedTokens for recently updated resources. By doing this, permissions-api becomes responsible for determining the freshness of lookups. On permissions checks for a given resource, the following consistency strategy is used:
+
+* If a ZedToken exists for that resource, use `at_least_as_fresh` with the given ZedToken
+* If no ZedToken exists, use `minimize_latency`
+
+The advantage of this approach is that it keeps management of consistency within permissions-api, meaning future changes do not (necessarily) require updates to other services, nor does it change current API semantics. Additionally, it does not introduce new dependencies to permissions-api, instead leveraging CRDB's availability and fault tolerance guarantees.
+
+## Constraints and limitations
+
+As designed, this only provides immediately visible updates for a given resource; other indirectly affected resources are not updated (but could be in the future). Additionally, given CRDB leaseholder behavior, multi-region deployments may see longer durations when performing permissions checks if the ZedToken table is not a global table.
+
+## Alternatives considered
```

Comment:
I'm still trying to understand this full stack, so excuse my ignorance. It seems like another alternative is no ZedTokens at all. Reading [this post](https://authzed.com/blog/prevent-newenemy-cockroachdb) seems to indicate that this won't happen in CRDB. Also it seems even less likely because of our [prefixing](https://github.com/infratographer/permissions-api/blob/main/docs/iapl.md#mapping-policies-to-spicedb).

Am I missing something in my reading of that post and our usage of SpiceDB?

Reply:
So that blog post is primarily about controlling write ordering in CRDB, not ACL updates. There are a few strategies we can adopt if the current default of using a single key for ordering transactions becomes a problem, but we haven't run into scaling issues there.

Where things get more interesting is around ZedTokens. The use of ZedTokens isn't really specific to CRDB, but rather inherent to how Zanzibar and SpiceDB operate. SpiceDB revisions (known as snapshots in the Zanzibar paper) are generally some number of seconds long and reads happen within a given revision. All datastores, including CRDB, use revisions for reading data.

ZedTokens allow for the client to request data newer than whatever a given SpiceDB replica might have available in its cache or whatever the current revision it's using is. In a CRDB deployment environment that uses follower reads, this would mean requesting data that might be newer than what the nearest replica has, thus incurring a one-time round trip to the leaseholder coordinating work. Following this, the data are cached in memory in SpiceDB for the remainder of the revision duration. Other storage engines don't have the same horizontal/multi-region scaling capability that CRDB offers.

Not using ZedTokens would require either waiting for a new revision to elapse before ACL updates are visible or leveraging [full consistency][consistency] in permissions checks, which means we get none of the caching benefits SpiceDB provides.

[consistency]: https://authzed.com/docs/spicedb/concepts/consistency#levels